Translated "render-props.md" (#120)
reactjs/de.reactjs.org: content/docs/render-props.md

}
```

Anstatt die `<Mouse>` Komponente einfach zu klonen und etwas Anderes in die `render` Methode spezifisch zu hard-coden, verwenden wir ein `render` Prop, welches `<Mouse>` dynamisch nutzen und bestimmen kann, was es rendert.

Genauer ausgedrückt, **ein „render prop“ ist ein Funktionsprop, dass eine Komponente nutzt, um zu wissen, was es rendern muss.**

Diese Technik macht das gewünschte Verhalten übertragbar. Um das Verhalten zu erhalten, rendern wir eine `<Mouse>` mit einem `render` Prop, welches mitteilt, was es zu rendern hat. Selbstverständlich unter Berücksichtung der aktuellen (x, y) Koordinaten des Cursors.

Eine interessante Sache über „render props“: Man kann die meisten „[higher-order components](/docs/higher-order-components.html)“ (HOC) durch normale Komponenten mit einem "render prop" implementieren. Wenn Sie beispielsweise ein `withMouse` HOC anstelle einer `<Mouse>` Komponente nutzen möchten, können Sie dies einfach mit einer normalen `<Mouse>` mit "render prop" erstellen: 

```js
// Wenn Sie wirklich ein HOC benötigen können Sie einfach

> // eines mit einer regulären Komponente mit "render prop" erstellen!

Reviewer: ```suggestion
// mit einer regulären Komponente mit einem "render prop" erstellen!
```

}
```

Ein render Prop zu nutzen hat somit auch den Vorteil, beide Strukturen nutzen zu können.

## Andere Props außer `render` nutzen {#using-props-other-than-render}

Es ist wichtig sich zu merken, dass nur weil diese Struktur „render props“ heißt, es *nicht zwingend notwendig ist ein Prop namens `render` zu nutzen, um auch diese Struktur zu nutzen*. Tatsächlich ist [*jedes* Prop, welches eine Funktion ist, die auch eine Komponente nutzen kann, um zu wissen was zu rendern ist, technisch ein „render prop“](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce).

> Auch wenn das Beispiel oben ein `render` nutzt, könnten wir auch einfach ein `children` Prop nutzen!

Reviewer: ```suggestion
Auch wenn das Beispiel oben `render` nutzt, könnten wir auch einfach ein `children`-Prop nutzen!
```

```js
<Mouse children={mouse => (

>   <p>Momentane Mausposition ist: {mouse.x}, {mouse.y}</p>

Reviewer: ```suggestion
  <p>Die momentane Mausposition ist: {mouse.x}, {mouse.y}</p>
```

)}/>
```

> Merke! Das `children` Prop muss nicht in den Attributen im „JSX element“ gelistet sein. Stattdessen können sie es direkt *in* das element einfügen!

Reviewer: ```suggestion
Und merke, dass das `children`-Prop nicht unbedingt in der Liste der "Attribute" im JSX-Element genannt werden muss. Stattdessen kannst du es direkt *in* das Element einfügen!
```

```js
<Mouse>
  {mouse => (

>     <p>Momentane Mausposition ist: {mouse.x}, {mouse.y}</p>

Reviewer: ```suggestion
    <p>Die momentane Mausposition ist: {mouse.x}, {mouse.y}</p>
```

  )}
</Mouse>
```

> Sie können sehen wie diese Technik in der [react-motion](https://github.com/chenglou/react-motion) API genutzt wird.

Reviewer: ```suggestion
Du kannst sehen wie diese Technik in der [react-motion](https://github.com/chenglou/react-motion) API genutzt wird.
```

> Da diese Technik ein wenig ungewöhnlich ist, möchten Sie wahrscheinlich explizit angeben das `children` eine Funktion in Ihren `propTypes` ist, wenn Sie eine API wie diese erstellen.

Reviewer: ```suggestion
Da diese Technik etwas ungewöhnlich ist, möchtest du wahrscheinlich explizit angeben, dass `children` eine Funktion in deinen `propTypes` ist, wenn du eine API wie diese erstellst.
```

```js
Mouse.propTypes = {
  children: PropTypes.func.isRequired
};
```

## Vorbehalte {#caveats}

> ### Seien Sie vorsichtig wenn Sie Render Props mit React.PureComponent nutzen {#be-careful-when-using-render-props-with-reactpurecomponent}

Reviewer: ```suggestion
### Sei vorsichtig wenn du render-Props mit React.PureComponent nutzt {#be-careful-when-using-render-props-with-reactpurecomponent}
```

> Bei der Nutzung eines render Props kann dieses die Vorteile negieren, welche vom [`React.PureComponent`](/docs/react-api.html#reactpurecomponent) kommen, sollten Sie die Funktion innerhalb der `render` Methode erstellen. Grund hierfür ist, dass der oberflächliche Prop Vergleich immer ein `false` für neue Props wiedergibt und jeder `render` in diesem Fall einen neuen Wert für den render Prop generiert.

Reviewer: ```suggestion
Bei der Nutzung eines render-Props kann dieses die Vorteile zunichte machen, welche vom [`React.PureComponent`](/docs/react-api.html#reactpurecomponent) kommen, solltest du die Funktion innerhalb der `render` Methode erstellen. Grund hierfür ist, dass der oberflächliche Prop Vergleich immer ein `false` für neue Props zurückgibt und jedes `render` in diesem Fall einen neuen Wert für das render-Prop generiert.
```

> Ein Beispiel: Sollten wir weiter mit unserer `<Mouse>` Komponente von oben arbeiten und `Mouse` erweitert, statt `React.Component`, `React.PureComponent` würde es folgendermaßen aussehen:

Reviewer: ```suggestion
Wenn wir zum Beispiel mit unserer `<Mouse>`-Komponente von oben fortfahren und `Mouse` statt `React.Component`, `React.PureComponent` nutzen würde, würde es folgendermaßen aussehen:
```

```js
class Mouse extends React.PureComponent {
  // Gleiche Implementierung wie oben...
}

class MouseTracker extends React.Component {
  render() {
    return (
      <div>
        <h1>Bewege die Maus!</h1>

        {/*

>           Nicht gut! Der Wert des `render` Prop wird

Reviewer: ```suggestion
          Nicht gut! Der Wert des `render`-Prop wird
```

>           bei jedem rendern verändert.

Reviewer: ```suggestion
          bei jedem Rendern verändert.
```

}
```

> In diesem Beispiel wird jedes Mal, wenn `<MouseTracker>` rendert, eine neue Funktion mit dem Wert vom `<Mouse render>` Prop generiert, wodurch der Effekt von der `<Mouse>` erweiterten `React.PureComponent` negiert wird!

Reviewer: ```suggestion
In diesem Beispiel wird jedes Mal, wenn `<MouseTracker>` rendert, eine neue Funktion mit dem Wert vom `<Mouse render>`-Prop generiert, wodurch der Effekt von der `<Mouse>` erweiterten `React.PureComponent` zunichte gemacht wird!
```

> Um dieses Problem zu umgehen, können Sie manchmal das Prop als eine Instanz-Methode definieren, in etwa so:

Reviewer: ```suggestion
Um dieses Problem zu umgehen, kannst du manchmal das Prop als eine Instanz-Methode definieren, in etwa so:
```

```js
class MouseTracker extends React.Component {
  // Als Instanz-Methode definiert wird `this.renderTheCat` immer
  // die Selbe Funktion referrieren, wenn sie gerendert wird
  renderTheCat(mouse) {
    return <Cat mouse={mouse} />;
  }

  render() {
    return (
      <div>
        <h1>Bewege die Maus!</h1>
        <Mouse render={this.renderTheCat} />
      </div>
    );
  }
}
```

> In Fällen, in denen Sie das Prop nicht statisch definieren können (bspw., weil Sie es über den Props der Komponenten und/oder dem Zustand schließen müssen), sollte `<Mouse>` den `React.Component` stattdessen erweiten.

Reviewer: ```suggestion
In Fällen, in denen du das Prop nicht statisch definieren kannst (bspw., weil du es über die Props der Komponenten und/oder dem State schließen musst), sollte `<Mouse>` den `React.Component` stattdessen erweitern.
```
